editor/src/clj/editor/breakpoints_view.clj

```diff
@@ -0,0 +1,558 @@
+    (prefs/set! prefs [:code :breakpoints] bps-prefs)))
+
+(defn- restore-breakpoints! [project prefs]
+  (g/with-auto-evaluation-context evaluation-context
+    (let [breakpoints (keep #(when-some [resource (workspace/find-resource (project/workspace project)
+                                                                           (:proj-path %)
+                                                                           evaluation-context)]
+                               (assoc % :resource resource))
+                            (prefs/get prefs [:code :breakpoints]))
+          script-bps (collect-script-nodes-from-breakpoints project breakpoints evaluation-context)]
+      (g/transact
+        (for [{:keys [script-node breakpoints]} script-bps
+              :let [updated-regions (update-script-regions-from-breakpoints script-node breakpoints evaluation-context)]]
+          (g/set-property script-node :regions updated-regions))))))
+
+(defn create-breakpoint-view-renderer [project localization prefs breakpoints-container open-resource-fn]
+  (restore-breakpoints! project prefs)
+  (let [open-res-fn (make-open-resource-fn open-resource-fn)]
+    (fx/mount-renderer
+      state
+      (fx/create-renderer
+       :error-handler #'error-reporting/report-exception!
+       :middleware (comp
+                     fxui/wrap-dedupe-desc
+                     (fx/wrap-map-desc
+                       (fn [state]
+                         {:fx/type fx/ext-watcher
+                          :ref localization
+                          :key :localization-state
+                          :desc {:fx/type breakpoints-view
+                                 :parent breakpoints-container
+                                 :state state}})))
+       :opts {:fx.opt/map-event-handler #(handle-breakpoint-event! project open-res-fn %)})))
+  (let [table (.lookup (ui/main-root) "#breakpoints-table-view")]
+    (ui/context! table :breakpoints-view
+                 {:project project :table table}
+                 (ui/->selection-provider table)
+                 {}
+                 {resource/Resource (fn [idx] (-> @state :breakpoints (get idx) :resource))})
+    (ui/register-context-menu table ::breakpoint-menu))
+  (let [timer (ui/->timer
+               5
+               "breakpoints-view-update-timer"
+               (fn [_ _ _]
+                 (when-not (ui/ui-disabled?)
+                   (let [breakpoints (g/node-value project :breakpoints)]
+                     (save-breakpoints! prefs breakpoints)
+                     (swap! state assoc :breakpoints breakpoints)))))]
+    (ui/timer-start! timer)))
+
+(handler/defhandler :breakpoints.toggle-selected-enabled :breakpoints-view
+  (run [project]
+    (g/with-auto-evaluation-context evaluation-context
+      (let [breakpoints (:breakpoints @state)
+            selected (map #(get breakpoints %) (:selected-indices @state))]
```

Comment:
Once you get rid of the global state, you'll be able to get breakpoints from the project, so there is a single source of truth for breakpoints (the graph).

`selected-indices` should be providable as a context, or maybe as selection/selection provider...

ALso please avoid lazy seqs.